File: java/src/main/java/deltix/luminary/FileDef.java
package deltix.luminary;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class FileDef implements TypeScope {
    private final String fileName;
    private final NamespaceDef namespace;
    private final List<DecoratorDef> definedDecorators = new ArrayList<>();
    private final List<ClassDef> definedClasses = new ArrayList<>();
    private final List<InterfaceDef> definedInterfaces = new ArrayList<>();
    private final List<EnumerationDef> definedEnumerations = new ArrayList<>();
    private final List<ImportDef> imports = new ArrayList<>();
    private final List<NameValuePair> options = new ArrayList<>();

    public FileDef(@NotNull NamespaceDef namespace, @Nullable String fileName) {
        this.namespace = namespace;
        this.fileName = fileName;

        if (fileName != null)
            namespace.getFiles().put(fileName, this);
    }

    @Nullable
    public String getFileName() {
        return fileName;
    }

    @NotNull
    public NamespaceDef getNamespace() {
        return namespace;
    }

    @NotNull
    public List<DecoratorDef> getDefinedDecorators() {
        return definedDecorators;
    }

    @NotNull
    public List<ClassDef> getDefinedClasses() {
        return definedClasses;
    }

    @NotNull
    public List<InterfaceDef> getDefinedInterfaces() {
        return definedInterfaces;
    }

    @NotNull
    public List<EnumerationDef> getDefinedEnumerations() {
        return definedEnumerations;
    }

    @NotNull
    public FileDef getFile() {
        return this;
    }

    public TypeScope getParent() {
        return null;
    }

    @NotNull
    public List<ImportDef> getImports() {
        return imports;
    }

    @NotNull
    public List<NameValuePair> getOptions() {
        return options;
    }
}

File: csharp/main/FileDef.cs
using System;
using System.Collections.Generic;
using System.IO;
using JetBrains.Annotations;

namespace Deltix.Luminary
{
	public class FileDef : ITypeScope
	{
		/// <summary>
		/// Creates an instance that describes a particular file within the give namespace.
		/// </summary>
		/// <param name="namespace">Instance of the <see cref="NamespaceDef"/> this file belongs to.</param>
		/// <param name="fileName">Name of the file.</param>
		public FileDef([NotNull] NamespaceDef @namespace, [NotNull] String fileName)
		{
			Namespace = @namespace;
			FileName = fileName;
			Namespace.Files.Add(fileName, this);
		}

		[NotNull]
		public ProjectDef Project => Namespace.Project;

		[NotNull]
		public NamespaceDef Namespace { get; }

		/// <summary>
		/// Name of the file.
		/// </summary>
		[NotNull]
		public String FileName { get; }

		[NotNull, ItemNotNull]
		public List<ImportDef> Imports { get; } = new List<ImportDef>();

		[NotNull, ItemNotNull]
		public List<NameValuePair> Options { get; } = new List<NameValuePair>();

		public List<DecoratorDef> DefinedDecorators { get; } = new List<DecoratorDef>();

		public List<ClassDef> DefinedClasses { get; } = new List<ClassDef>();

		public List<EnumerationDef> DefinedEnumerations { get; } = new List<EnumerationDef>();

		public List<InterfaceDef> DefinedInterfaces { get; } = new List<InterfaceDef>();

		public Dictionary<String, TypeCustom> DefinedTypes { get; } = new Dictionary<String, TypeCustom>();

		[CanBeNull]
		public ITypeScope Parent => null;

		[NotNull]
		public FileDef File => this;

		/// <summary>
		/// Returns the path to this file on disk relative to the location of the project file.
		/// </summary>
		/// <returns>Relative path to this file on disk.</returns>
		[NotNull]
		public String GetPath()
		{
			// TODO: Should be Path.AltDirectorySeparatorChar or something similar.
			return Namespace.Namespace.Replace('.', '/') + FileName;
		}
	}
}
